src/core/components/pin.js

```diff
@@ -0,0 +1,412 @@
+'use strict'
+
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const CID = require('cids')
+const pinSet = require('./pin-set')
+const resolvePath = require('../utils').resolvePath
+const promisify = require('promisify-es6')
+const multihashes = require('multihashes')
+const Key = require('interface-datastore').Key
+const eachLimit = require('async/eachLimit')
+const series = require('async/series')
+const waterfall = require('async/waterfall')
+const parallel = require('async/parallel')
+const someLimit = require('async/someLimit')
+const once = require('once')
+
+// arbitrary limit to number of concurrent dag operations
+const concurrencyLimit = 300
+
+function toB58String (hash) {
+  return new CID(hash).toBaseEncodedString()
+}
+
+module.exports = function pin (self) {
+  let directPins = new Set()
+  let recursivePins = new Set()
+
+  const pinDataStoreKey = new Key('/local/pins')
+
+  const repo = self._repo
+  const dag = self.dag
+
+  const pin = {
+    types: {
+      direct: 'direct',
+      recursive: 'recursive',
+      indirect: 'indirect',
+      all: 'all'
+    },
+
+    clear: () => {
+      directPins.clear()
+      recursivePins.clear()
+    },
+
+    set: pinSet(dag),
+
+    add: promisify((paths, options, callback) => {
+      if (typeof options === 'function') {
+        callback = options
+        options = null
+      }
+      callback = once(callback)
+      const recursive = options ? options.recursive : true
+
+      resolvePath(self.object, paths, (err, mhs) => {
+        if (err) { return callback(err) }
+
+        // verify that each hash can be pinned
+        parallel(mhs.map(multihash => cb => {
```

Comment:
Super minor, but `async.map` will save you creating an function for each multihash...